Revert improved filtering for variations

Konamiman opened the pull request:
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This pull request reverts the functionality introduced in #26260 (later refined by #27175, #27190, #27508) in which filtering by
attribute using the layered nav widget was improved to handle the cases of variations out of stock. The revert is a response to the numerous problems related to filtering reported by users in WooCommerce 4.4 and 4.5, including two more that we hadn't detected previously:

* Attributes not used for variations aren't handled correctly.
* Products marked as "hidden for catalog" are still being counted.

We'll figure out a better way to achieve the same functionality and re-implement the feature in a future version of WooCommerce.

Not all the code has been reverted, only the code that resulted in visible functionality changes. Thus, the code that generates
term relationships for variations is still in place to keep database consistency and to keep the reverting changes to the minimum needed.

May close (hopefully) https://github.com/woocommerce/woocommerce/issues/27419.

**Note for reviewers:** To check the code changes in the widget itself you may find it useful to open [the diff of this PR against 4.3.0](https://github.com/woocommerce/woocommerce/compare/4.3.0...revert-improved-filtering-for-variations), go to "Files changed" and search for `includes/widgets/class-wc-widget-layered-nav.php` (I didn't fully revert this one since there are small improvements that can be kept, e.g. for the code sniffer).

### How to test the changes in this Pull Request:

Basically, verify that the filtering widget works as in WooCommerce 4.3. This implies that even if "hide out of stock items" is set, there are simple and variable products, and at least one of them has stock, the widget will count all the attributes including:

* The ones used by simple products.
* The ones used by variable products, but not for variations.
* The ones used by variable products, for variations, even those not having stock.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Revert the changes in filtering by attribute that were introduced in WooCommerce 4.4

Konamiman committed:
Revert improved filtering for variations.

This commit reverts the functionality introduced in PR #26260
(later refined by #27175, #27190, #27508) in which filtering by
attribute using the layered nav widget was improved to handle the
cases of variations out of stock. The revert is a response to the
numerous problems reported by users in Woo 4.4 and 4.5

Not all the code has been reverted, only the code that resulted in
visible functionality changes. Thus, the code that generates
term relationships for variations is still in place to keep database
consistency and to keep the reverting changes to the minimum needed.

ObliviousHarmony reviewed (approved):
LGTM!

claudiosanches commented on templates/content-product.php:
```diff
@@ -20,7 +20,7 @@
 global $product;
 
 // Ensure visibility.
-if ( empty( $product ) || false === wc_get_loop_product_visibility( $product->get_id() ) || ! $product->is_visible() ) {
```
We should change the `wc_get_loop_product_visibility()` function to make sure it doesn't runs `wc_set_loop_product_visibility()` anymore.
Note that some people may have replaced those templates in their themes.

Konamiman replied:
Not sure if I understand what you mean. `wc_get_loop_product_visibility` is just doing a call to `wc_get_loop_prop`, or do you mean something else?

vedanshujain commented:
I am restoring all the related branches for review, will delete them later myself (so please don't delete them now :) ).

vedanshujain commented on includes/widgets/class-wc-widget-layered-nav.php:
```diff
@@ -344,93 +344,51 @@ protected function layered_nav_dropdown( $terms, $taxonomy, $query_type ) {
+		$term_ids_sql   = '(' . implode( ',', array_map( 'absint', $term_ids ) ) . ')';
 
 		// Generate the first part of the query.
 		// This one will return non-variable products and variable products with concrete values for the attributes.
```
no need for this comment anymore

vedanshujain reviewed (changes requested):
Great work! I have added some comments inline, and some suggestions in the summary comments (which are optional).

Improvements we should do soon (these would have been blockers if we were not doing it in a fix release)

- [ ] Update function `wc_update_440_insert_attribute_terms_for_variable_products` does too much for running in a 30s cron. The time it will take is mostly dependent on how many variations are there for products on the page. We should instead loop on variations (as opposed to their parent product), and instead of the page, we should fetch on ID and specify explicit order. Fetching using the page is prone to race conditions and filters interfering with the query so we should make it as explicit as possible.

Suggestions (not blockers, can be done in a different PR) 
- [ ] When [deleting associated taxanomies](https://github.com/woocommerce/woocommerce/blob/d48f1d4e2efcfab8206799df2085f021ec273722/includes/data-stores/class-wc-product-variable-data-store-cpt.php#L664), fetch the term_ids directly and use `wp_remove_object_terms`. Right now we are effectively looping across all the product attributes which is going to be very slow for large shops.
- [ ] Same with the block add to [sync attribute to terms table](https://github.com/woocommerce/woocommerce/blob/d48f1d4e2efcfab8206799df2085f021ec273722/includes/data-stores/class-wc-product-variation-data-store-cpt.php#L499) as well. We should find a way to avoid calling `wc_get_attribute_taxonomy_names` and instead fetch the existing attributes and delete them directly.
- [ ] Same in [product variation delete](https://github.com/woocommerce/woocommerce/blob/d48f1d4e2efcfab8206799df2085f021ec273722/includes/class-wc-product-variation.php#L601).

Konamiman commented:
> Update function wc_update_440_insert_attribute_terms_for_variable_products does too much for running in a 30s cron.

This migration is not needed once the revert is in place, since these extra entries in the term relationships table won't be used for anything. I didn't remove it to keep the revert as simple as possible, but we'll probably be able to remove this migration (and the already existing entries) later.

And the same for the other suggestions about syncing variation terms.

Konamiman commented on includes/class-wc-query.php:
```diff
@@ -395,52 +382,6 @@ public function remove_product_query_filters( $posts ) {
-	 *
-	 * @return int Adjusted posts count.
-	 */
-	public function adjust_posts_count( $count, $query ) {
```
Ok. Note however that we'll probably need to reintroduce the same function later when we re-implement the thing, so maybe it's better to just make it return the count unchanged as you say, but without deprecating it?

Konamiman committed:
Reintroduce adjust_posts_count, and remove unnecessary comment.

We can't remove that method since it's public, but its body has
been changed so that it just returns the input unchanged.

vedanshujain reviewed (approved):
Approving the PR with pending improvements in order to keep the changeset as small as possible for the fix release.
